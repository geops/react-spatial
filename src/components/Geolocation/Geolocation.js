import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import OLMap from 'ol/Map';
import { transform } from 'ol/proj';
import Point from 'ol/geom/Point';
import Feature from 'ol/Feature';

import Style from 'ol/style/Style';
import Circle from 'ol/style/Circle';
import Fill from 'ol/style/Fill';
import Stroke from 'ol/style/Stroke';

import VectorLayer from 'ol/layer/Vector';
import VectorSource from 'ol/source/Vector';

import { FaRegDotCircle } from 'react-icons/fa';

const propTypes = {
  /**
   * CSS class of the button.
   */
  className: PropTypes.string,

  /**
   *  Children content of the Geolocation button.
   */
  children: PropTypes.node,

  /**
   * Map.
   */
  map: PropTypes.instanceOf(OLMap).isRequired,

  /**
   * Function triggered when geolocating is not successful.
   */
  onError: PropTypes.func,

  /**
   * If true, the map is not centered after it has been dragged once.
   */
  noCenterAfterDrag: PropTypes.bool,

  /**
   * Color (Number array with rgb values) or style function.
   * If a color is given, the style is animated.
   */
  colorOrStyleFunc: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.number),
    PropTypes.func,
  ]),
};

const defaultProps = {
  className: 'rs-geolocation',
  children: <FaRegDotCircle focusable={false} />,
  onError: () => {},
  noCenterAfterDrag: false,
  colorOrStyleFunc: [235, 0, 0],
};

/**
 * This component displays a geolocator.
 */
class Geolocation extends PureComponent {
  constructor(props) {
    super(props);
    const { map, noCenterAfterDrag } = this.props;

    this.layer = new VectorLayer({
      source: new VectorSource(),
    });

    this.isCentered = true;
    if (noCenterAfterDrag) {
      map.on('pointerdrag', () => {
        this.isCentered = false;
      });
    }

    this.state = {
      active: false,
    };
  }

  toggle() {
    const { active } = this.state;
    const { onError } = this.props;
    const geolocation = 'geolocation' in navigator;

    if (!geolocation) {
      onError();
    } else if (!active) {
      this.watch = navigator.geolocation.watchPosition(
        this.activate.bind(this),
        this.error.bind(this),
        {
          enableHighAccuracy: true,
        },
      );
    } else {
      this.deactivate();
    }
  }

  error() {
    const { onError } = this.props;

    this.deactivate();
    onError();
  }

  deactivate() {
    window.clearInterval(this.interval);
    this.layer.setMap(null);
    navigator.geolocation.clearWatch(this.watch);

    this.setState({
      active: false,
    });
  }

  activate(position) {
    const { map } = this.props;

    const code = map
      .getView()
      .getProjection()
      .getCode();
    const pos = transform(
      [position.coords.longitude, position.coords.latitude],
      'EPSG:4326',
      code,
    );

    const point = new Point(pos);
    this.highlight(point);
    this.layer.setMap(map);
    if (this.isCentered) {
      map.getView().setCenter(pos);
    }

    this.setState({
      active: true,
    });
  }

  highlight(point) {
    const { colorOrStyleFunc } = this.props;

    let decrease = true;
    let opacity = 0.5;
    let rotation = 0;
    let feature;

    window.clearInterval(this.interval);
    this.interval = window.setInterval(() => {
      rotation += 0.03;
      decrease = opacity < 0.1 ? false : decrease;
      decrease = opacity > 0.5 ? true : decrease;
      opacity += decrease ? -0.03 : 0.03;
      if (feature) {
        feature.changed();
      }
    }, 50);

    feature = new Feature({
      geometry: point,
      source: new VectorSource(),
    });

    if (Array.isArray(colorOrStyleFunc)) {
      const color = colorOrStyleFunc;

      feature.setStyle(() => {
        const circleStyle = new Style({
          image: new Circle({
            radius: 20,
            rotation,
            fill: new Fill({
              color: 'rgba(255, 255, 255, 0.01)',
            }),
            stroke: new Stroke({
              lineDash: [30, 10],
              width: 6,
              color: `rgba(${color[0]}, ${color[1]}, ${color[2]}, ${opacity})`,
            }),
          }),
        });

        circleStyle.getImage().setRotation(rotation);

        return [
          new Style({
            image: new Circle({
              radius: 10,
              fill: new Fill({
                color: `rgba(${color[0]}, ${color[1]}, ${color[2]}, 0.5)`,
              }),
            }),
          }),
          circleStyle,
        ];
      });
    } else {
      feature.setStyle(colorOrStyleFunc);
    }

    this.layer.getSource().clear();
    this.layer.getSource().addFeature(feature);
  }

  render() {
    const { children, className } = this.props;
    // Remove component props from other HTML props.
    const other = Object.entries(this.props).reduce((props, [key, value]) => {
      return propTypes[key] ? props : { ...props, [key]: value };
    }, {});
    const { active } = this.state;

    return (
      <div
        role="button"
        tabIndex="0"
        className={`${className} ${active ? 'rs-active' : ''}`}
        onClick={() => this.toggle()}
        onKeyPress={e => e.which === 13 && this.toggle()}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...other}
      >
        {children}
      </div>
    );
  }
}

Geolocation.propTypes = propTypes;
Geolocation.defaultProps = defaultProps;

export default Geolocation;
