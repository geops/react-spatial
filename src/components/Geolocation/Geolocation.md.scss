.tm-geolocation-example {
  position: relative;

  .tm-geolocation {
    top: 10px;
    right: 10px;
    position: absolute;
  }
}
