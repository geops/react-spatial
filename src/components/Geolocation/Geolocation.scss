.tm-geolocation {
  cursor: pointer;
  height: $btn-size-base;
  width: $btn-size-base;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  text-align: center;
  font-size: 20px;
  color: white;
  background-color: $btn-primary-color;
  transition: background-color 0.5s ease;

  &:hover {
    background-color: $btn-primary-color-hover;
  }

  &.blink {
    animation-name: blinking;
    animation-duration: 3s;
    animation-timing-function: linear;
    animation-iteration-count: infinite;
  }

  @keyframes blinking {
    0% { color: white; }
    50% { color: $btn-primary-color; }
    100% { color: white; }
  }

  /**
   * Loading animation
   */
  @keyframes &.blink {
    /**
     * At the start of the animation the dot
     * has an opacity of 0.2
     */
    0% {
      opacity: 0.2;
    }

    /**
     * At 20% the dot is fully visible and
     * then fades out slowly
     */
    20% {
      opacity: 1;
    }

    /**
     * Until it reaches an opacity of 0.2 and
     * the animation can start again
     */
    100% {
      opacity: 0.2;
    }
  }
}
