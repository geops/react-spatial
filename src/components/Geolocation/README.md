#

This demonstrates the use of Geolocation.

```jsx
import React from 'react';
import Geolocation from 'react-spatial/components/Geolocation';
import BasicMap from 'react-spatial/components/BasicMap';
import OLMap from 'ol/Map';
import ConfigReader from 'react-spatial/ConfigReader';

class GeolocationExample extends React.Component {
  constructor(props) {
    super(props);

    this.center = [-10997148, 4569099];
    this.map = new OLMap({ controls: [] });
  }

  componentDidMount() {
    const layers = ConfigReader.readConfig(
      this.map,
      treeData,
    );
  }

  render() {
    return (
      <div className="tm-geolocation-example">
        <BasicMap
          map={this.map}
          center={this.center}
          zoom={3}
          layers={this.layers}
        />
        <Geolocation
          title='Lokalisieren'
          map={this.map}
          noCenterAfterDrag
          onError={() => window.alert('Geolokalisierung war nicht erfolgreich')}
          colorOrStyleFunc={[0, 61, 133]}
        />
      </div>
    );
  }
}

<GeolocationExample />;
```
