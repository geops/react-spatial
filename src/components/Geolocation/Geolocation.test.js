import React from 'react';
import renderer from 'react-test-renderer';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import 'jest-canvas-mock';
import Map from 'ol/Map';
import View from 'ol/View';
import MapEvent from 'ol/MapEvent';
import Geolocation from './Geolocation';

configure({ adapter: new Adapter() });

const geolocationBackup = global.navigator.geolocation;

const mockGeolocation = () => {
  const mock = {
    clearWatch: jest.fn(),
    getCurrentPosition: jest.fn(),
    watchPosition: onSuccess => {
      onSuccess({
        coords: {
          latitude: 47.9913611,
          longitude: 7.84868,
          accuracy: 55,
        },
        timestamp: 1552660077044,
      });
    },
  };

  global.navigator.geolocation = mock;
};

const mockMissingGeolocation = () => {
  delete global.navigator.geolocation;
};

const restoreGeolocation = () => {
  global.navigator.geolocation = geolocationBackup;
};

describe('Geolocation', () => {
  let map;

  beforeEach(() => {
    const target = document.createElement('div');
    const { style } = target;
    style.position = 'absolute';
    style.left = '-1000px';
    style.top = '-1000px';
    style.width = '100px';
    style.height = '100px';
    document.body.appendChild(target);

    map = new Map({
      target,
      view: new View({
        center: [0, 0],
        resolutions: [1],
        zoom: 0,
      }),
    });
    map.renderSync();
  });

  afterEach(() => {
    const target = map.getTarget();
    map.setTarget(null);
    if (target && target.parentNode) {
      target.parentNode.removeChild(target);
    }
    map.dispose();
  });

  describe('should match snapshot', () => {
    test('minimum props', () => {
      const component = renderer.create(<Geolocation map={map} />);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('with title', () => {
      const component = renderer.create(
        <Geolocation map={map} title="Lokalisieren" />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('with class name', () => {
      const component = renderer.create(
        <Geolocation map={map} className="my-class-name" />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  test('should use children', () => {
    mockGeolocation();

    const wrapper = mount(<Geolocation map={map}>test</Geolocation>);

    const text = wrapper
      .find('.rs-geolocation')
      .first()
      .text();

    expect(text).toBe('test');

    restoreGeolocation();
  });

  describe('button classes', () => {
    test('class should be active', () => {
      mockGeolocation();

      const wrapper = mount(<Geolocation map={map} />);
      const basic = wrapper.getDOMNode();

      wrapper
        .find('.rs-geolocation')
        .first()
        .simulate('click');

      expect(basic.className).toBe('rs-geolocation rs-active');

      restoreGeolocation();
    });

    test('class should not be active', () => {
      mockGeolocation();

      const wrapper = mount(<Geolocation map={map} />);
      const basic = wrapper.getDOMNode();

      wrapper
        .find('.rs-geolocation')
        .first()
        .simulate('click')
        .simulate('click');

      expect(basic.className).toBe('rs-geolocation ');

      restoreGeolocation();
    });
  });

  test(`highlight on first toggle`, () => {
    mockGeolocation();

    const component = shallow(<Geolocation map={map} />);
    const instance = component.instance();
    const spy = jest.spyOn(instance, 'highlight');
    instance.toggle();
    expect(spy).toHaveBeenCalled();

    restoreGeolocation();
  });

  test(`error function should be called`, () => {
    mockMissingGeolocation();

    class ErrorHandler {
      static onError() {}
    }

    const spy = jest.spyOn(ErrorHandler, 'onError');

    const wrapper = mount(
      <Geolocation map={map} onError={() => ErrorHandler.onError()} />,
    );

    wrapper
      .find('.rs-geolocation')
      .first()
      .simulate('click');

    expect(spy).toHaveBeenCalled();

    restoreGeolocation();
  });

  describe('map centering', () => {
    test('centers map', () => {
      mockGeolocation();

      const center1 = [742952.8821531708, 6330118.608483334];
      map.getView().setCenter(center1);

      const component = shallow(<Geolocation map={map} />);
      component.instance().toggle();

      const center2 = map.getView().getCenter();
      expect(center1).not.toEqual(center2);

      restoreGeolocation();
    });

    test('no center after drag', () => {
      mockGeolocation();

      const center1 = [742952.8821531708, 6330118.608483334];
      map.getView().setCenter(center1);

      const component = shallow(<Geolocation map={map} noCenterAfterDrag />);
      map.dispatchEvent(new MapEvent('pointerdrag', map));
      component.instance().toggle();

      const center2 = map.getView().getCenter();
      expect(center1).toEqual(center2);

      restoreGeolocation();
    });
  });

  test('custom style function', () => {
    mockGeolocation();

    const styleFunc = jest.fn();

    const component = shallow(
      <Geolocation map={map} colorOrStyleFunc={styleFunc} />,
    );
    const instance = component.instance();
    instance.toggle();
    const style = instance.layer
      .getSource()
      .getFeatures()[0]
      .getStyle();

    expect(style).toBe(styleFunc);

    restoreGeolocation();
  });
});
