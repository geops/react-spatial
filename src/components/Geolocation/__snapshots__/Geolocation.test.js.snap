// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geolocation should match snapshot minimum props 1`] = `
<div
  className="tm-geolocation "
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <svg
    fill="currentColor"
    focusable={false}
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 56c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m0-48C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 168c-44.183 0-80 35.817-80 80s35.817 80 80 80 80-35.817 80-80-35.817-80-80-80z"
    />
  </svg>
</div>
`;

exports[`Geolocation should match snapshot with class name 1`] = `
<div
  className="my-class-name "
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <svg
    fill="currentColor"
    focusable={false}
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 56c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m0-48C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 168c-44.183 0-80 35.817-80 80s35.817 80 80 80 80-35.817 80-80-35.817-80-80-80z"
    />
  </svg>
</div>
`;

exports[`Geolocation should match snapshot with title 1`] = `
<div
  aria-label="Lokalisieren"
  className="tm-geolocation "
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
  title="Lokalisieren"
>
  <svg
    fill="currentColor"
    focusable={false}
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 56c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m0-48C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 168c-44.183 0-80 35.817-80 80s35.817 80 80 80 80-35.817 80-80-35.817-80-80-80z"
    />
  </svg>
</div>
`;
